torchtune/datasets/multimodal/_the_cauldron.py

```diff
@@ -0,0 +1,197 @@
+        new_system_prompt: Optional[str] = None,
+    ):
+        self.train_on_input = train_on_input
+        self.new_system_prompt = new_system_prompt
+        if column_map is not None:
+            if "images" not in column_map:
+                raise ValueError(
+                    "column_map must map 'images' to your expected column name if specified"
+                )
+            if "texts" not in column_map:
+                raise ValueError(
+                    "column_map must map 'texts' to your expected column name if specified"
+                )
+            self._column_map = column_map
+        else:
+            self._column_map = {"texts": "texts", "images": "images"}
+
+    def __call__(self, sample: Mapping[str, Any]) -> Mapping[str, Any]:
+        messages = []
+        for message in sample[self._column_map["texts"]]:
+            messages.append(
+                Message(
+                    role="user",
+                    content=[
+                        {"type": "image"},
+                        {"type": "text", "content": message["user"]},
+                    ],
+                    masked=not self.train_on_input,
+                )
+            )
+            messages.append(
+                Message(
+                    role="assistant",
+                    content=[{"type": "text", "content": message["assistant"]}],
+                )
+            )
+
+        if self.new_system_prompt is not None:
+            messages = [
+                Message(
+                    role="system", content=self.new_system_prompt, masked=True, eot=True
+                )
+            ] + messages
+
+        return {"messages": messages, "images": sample[self._column_map["images"]]}
+
+
+# TODO: point to Flamingo model transform as an example
+def the_cauldron_dataset(
+    model_transform: Transform,
+    *,
+    subset: str,
+    source: str = "HuggingFaceM4/the_cauldron",
+    column_map: Optional[Dict[str, str]] = None,
+    new_system_prompt: Optional[str] = None,
+    train_on_input: bool = True,
+    packed: bool = False,
+    split: str = "train",
+    **load_dataset_kwargs: Dict[str, Any],
+) -> Union[SFTDataset, PackedDataset]:
+    """
+    Support for family of image + text datasets similar to
+    `The Cauldron<https://huggingface.co/datasets/HuggingFaceM4/the_cauldron>_`
+    from Hugging Face Datasets.
+
+    The Cauldron consists of numerous datasets. You must specify one of the datasets
+    using the ``subset`` argument.
+
+    Args:
+        model_transform (Transform): model-specific transform class that takes in a sample dict and applies custom
+            transforms on the keys. It should consist of at minimum two components: text tokenization (called
+            on the "messages" field) and image transform (called on the "images" field). The keys returned by
+            the model transform should be aligned with the expected inputs into the model.
```

Comment:
Sorry to harp on this, but can we have a more explicit example until the Flamingo transform is available? Even if it's just a mock tokenizer and image transform with clearly-defined API contracts in a code snippet, I think that will help a lot